c_src/ep_decoder.c

```diff
@@ -887,7 +887,7 @@ fill_default(ErlNifEnv *env, ep_spot_t *spot)
             case field_sfixed32:
             case field_sfixed64:
             case field_enum:
-                *t++ = state->integer_zero;
+                *t++ = enif_make_int(env, 0);
```

Comment:
should reference to state be removed on line 894, 904 and 908 also?

Reply:
good catch - line 904 should be changed. i think the atoms are fine though, they haven't changed them upstream

Reply:
what if state is null by the time you come here? Also, what is an atom? You are referring to a variable in state, correct? is integer_zero a pointer?

Reply:
integer_zero, atom_false, and atom_undefined are all `ERL_NIF_TERM`s that are supposed to get initalized when the state object is created. (the state object is created when the NIF module is initalized.) We shouldn't be able to get to the point where we run this function at all if the state object is a null pointer.

`ERL_NIF_TERMS` represent erlang objects and are intended to be opaque. My guess for what's going on here is that the objects for zero as an integer and double represent some sort of pointer or reference that gets garbage-collected after we initalize the state, causing them to hang, but the atoms are likely just stored as indices or hashes in the atom table and never get hit by the GC.